File: 2/2_cuatrimestre/PDOO/practicas/practica 3/deepSpaceRuby/Weapon.rb
require_relative 'WeaponType'
require_relative 'WeaponToUI'

module Deepspace

    class Weapon

        def initialize (n, t, u)
            @name = n
            @type = t
            @uses = u        
        end

        def self.newCopy(w)
            new(w.name, w.type, w.uses)
        end

        def getUIversion
            return WeaponToUI.new(self)
        end

        def type
            @type
        end

        def name
            @name
        end
        
        def uses
            @uses
        end
        
        def power
            @type.power
        end

        def useIt
            if (@uses > 0)
                @uses -= 1
                power
            else
                1.0
            end
        end

        def to_s
            return getUIversion.to_s
        end

    end

end

File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpace_4/src/deepspace/Weapon.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

/**
 *
 * @author manbolq
 */
public class Weapon implements CombatElement{
    private String name;
    private WeaponType type;
    private int uses;
    
    
    /**************************************************************************/
    /*************************    CONSTRUCTORES   *****************************/
    /**************************************************************************/
    
    Weapon(String name, WeaponType type, int uses){
        this.name = name;
        this.type = type;
        this.uses = uses;
    }
    
    Weapon(Weapon w){
        name = w.name;
        type = w.type;
        uses = w.uses;
    }
    
    
    /**************************************************************************/
    /**************************    CONSULTORES   ******************************/
    /**************************************************************************/
    
    WeaponType getType(){
        return type;
    }
    
    @Override
    public int getUses(){
        return uses;
    }
    
    
    /**************************************************************************/
    /****************************    METODOS   ********************************/
    /**************************************************************************/
    
    public float power(){
        return type.getPower();
    }
    
    @Override
    public float useIt(){
        if (uses > 0){
            uses--;
            return power();
        }
        else return 1.0f;
    }
    
    
    @Override
    public String toString(){
        return getUIversion().toString();
    }
    
    
    WeaponToUI getUIversion(){
        return new WeaponToUI(this);
    }
}